### PR TITLE
Pv2 824 replace the commitments table with the new apprenticeship schema

### PR DESCRIPTION
**Pull request by samuel-dev-dba**

### CODE REVIEW

File: src/SFA.DAS.Payments.Database/zzz.Script.PostDeployment.sql [hunk 1]
```diff
@@ -51,6 +51,24 @@ IF NOT EXISTS (SELECT * FROM [Payments2].[JobType]  WHERE [Id] = 4)
...
+MERGE INTO [Payments2].[ApprenticeshipStatus]	 AS Target
+USING (VALUES
+(1	, N'Active'),
+(2	, N'Paused'),
+(3 , N'Stopped'),
+(4 , N'InActive')
+) AS Source ([Id],[Description])
+ON (Target.[Id] = Source.[Id])
+WHEN MATCHED AND
+  ( NULLIF(Source.[Description], Target.[Description]) IS NOT NULL OR NULLIF(Target.[Description], Source.[Description]) IS NOT NULL) THEN
```
Comment (kuzinator) [inline src/SFA.DAS.Payments.Database/zzz.Script.PostDeployment.sql]: Second part (after OR) is redundant, please remove

File: src/SFA.DAS.Payments.Model.Core/Entities/ApprenticeshipModel.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        public ApprenticeshipPaymentStatus? Status { get; set; }
+        public List<ApprenticeshipPriceEpisodeModel> ApprenticeshipPriceEpisodes { get; set; }
+    }
+
+    public  enum ApprenticeshipPaymentStatus:byte
+    {
+        Active = 1,
+        Paused = 2,
+        Stopped = 3,
+        InActive = 4
```
Comment (kuzinator) [inline src/SFA.DAS.Payments.Model.Core/Entities/ApprenticeshipModel.cs]: Please make "A" lower case

File: src/SFA.DAS.Payments.Model.Core/Entities/ApprenticeshipPriceEpisodeModel.cs [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+{
+    public class ApprenticeshipPriceEpisodeModel
+    {
+        public long Id { get; set; }
+        public long ApprenticeshipId { get; set; }
+        public DateTime StartDate { get; set; }
+        public DateTime? EndDate { get; set; }
+        public decimal Cost { get; set; }
+        public bool Removed { get; set; }
+        public ApprenticeshipModel Apprenticeship { get; set; }
```
Comment (kuzinator) [inline src/SFA.DAS.Payments.Model.Core/Entities/ApprenticeshipPriceEpisodeModel.cs]: Not sure about this. Do we need a circular dependency here?

File: src/SFA.DAS.Payments.Application/Repositories/PaymentsDataContext.cs [hunk 4]
```diff
@@ -21,7 +22,8 @@ protected override void OnModelCreating(ModelBuilder modelBuilder)
             base.OnModelCreating(modelBuilder);
             modelBuilder.HasDefaultSchema("Payments2");
             modelBuilder.ApplyConfiguration(new PaymentModelConfiguration());
-            modelBuilder.ApplyConfiguration(new CommitmentModelConfiguration());
+            modelBuilder.ApplyConfiguration(new ApprenticeshipModelConfiguration());
+            modelBuilder.ApplyConfiguration(new ApprenticeshipPriceEpisodeModelConfiguration());
```
Comment (kuzinator) [inline src/SFA.DAS.Payments.Application/Repositories/PaymentsDataContext.cs]: The schema has ApprenticeshipPause entity in it, is this not part of this story?

File: src/SFA.DAS.Payments.Database/Payments2/Tables/Payments2.Apprenticeship.sql [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+	EstimatedStartDate Date NOT NULL,
+	EstimatedEndDate Date NOT NULL,
+    [Priority]      INT  NOT NULL,
+	StandardCode BIGINT NULL,
+	ProgrammeType INT NULL,
+	FrameworkCode INT NULL,
+	PathwayCode INT NULL,
+	LegalEntityName NVARCHAR (100) NULL,
+	TransferSendingEmployerAccountId BIGINT NULL,
+	StopDate DateTime2 NULL,
```
Comment (kuzinator) [inline src/SFA.DAS.Payments.Database/Payments2/Tables/Payments2.Apprenticeship.sql]: Make this of type Date

File: src/SFA.DAS.Payments.Database/Payments2/Tables/Payments2.Apprenticeship.sql [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+	EstimatedEndDate Date NOT NULL,
+    [Priority]      INT  NOT NULL,
+	StandardCode BIGINT NULL,
+	ProgrammeType INT NULL,
+	FrameworkCode INT NULL,
+	PathwayCode INT NULL,
+	LegalEntityName NVARCHAR (100) NULL,
+	TransferSendingEmployerAccountId BIGINT NULL,
+	StopDate DateTime2 NULL, 
+    [Status] TINYINT NULL
```
Comment (kuzinator) [inline src/SFA.DAS.Payments.Database/Payments2/Tables/Payments2.Apprenticeship.sql]: What would a NULL mean? Can this be made non-nullable?

File: src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/Steps/EndToEndStepsBase.cs [hunk 7]
```diff
@@ -177,67 +177,90 @@ protected void AddTestAims(IList<Aim> aims, long ukprn)
...
-                await DataContext.Commitment.AddAsync(mappedCommitment);
+            }
+            apprenticeshipSpec.AccountId = TestSession.GetEmployer(apprenticeshipSpec.Employer).AccountId;
+            apprenticeshipSpec.Uln = TestSession.GetLearner(apprenticeshipSpec.Ukprn, apprenticeshipSpec.LearnerId).Uln;
+
+            ApprenticeshipPaymentStatus? apprenticeshipStatus = null;
+
+            switch (apprenticeshipSpec.Status?.ToLower())
+            {
+                case "active":
```
Comment (kuzinator) [inline src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/Steps/EndToEndStepsBase.cs]: This needs to default to "active"

File: src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/SFA.DAS.Payments.AcceptanceTests.EndToEnd.csproj [hunk 8]
```diff
@@ -920,6 +913,10 @@
       <Generator>SpecFlowSingleFileGenerator</Generator>
       <LastGenOutput>Levy learner submits Ilr several months into learning - PV2-340.feature.cs</LastGenOutput>
     </None>
+    <None Include="Tests\Learning Support\PV2-734 - Levy learner requires learning support doing an apprenticeship standard.feature">
```
[the remark this answers is not in the document]
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/SFA.DAS.Payments.AcceptanceTests.EndToEnd.csproj]: Not supposed to be there was doing look ahead task will fix

File: src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/SFA.DAS.Payments.AcceptanceTests.EndToEnd.csproj [hunk 9]
```diff
@@ -421,21 +419,11 @@
...
-      <AutoGen>True</AutoGen>
-      <DesignTime>True</DesignTime>
-      <DependentUpon>Levy learner changes employer with change to negotiated price at the mid of month with gap- PV2-364.feature</DependentUpon>
-    </Compile>
     <Compile Include="Tests\Employer Change of Circumstances\Levy learner changes employer with change to negotiated price at the mid of month- PV2-363.feature.cs">
       <AutoGen>True</AutoGen>
       <DesignTime>True</DesignTime>
       <DependentUpon>Levy learner changes employer with change to negotiated price at the mid of month- PV2-363.feature</DependentUpon>
     </Compile>
-    <Compile Include="Tests\Employer Change of Circumstances\Non-Levy learner changes to Levy employer Levy available- PV2-366.feature.cs">
```
[the remark this answers is not in the document]
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/SFA.DAS.Payments.AcceptanceTests.EndToEnd.csproj]: Noop the .cs file don't need to be in the solution  just run custom tool to generate a new one from spec

File: src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/SFA.DAS.Payments.AcceptanceTests.EndToEnd.csproj [hunk 10]
```diff
@@ -421,21 +419,11 @@
       <DesignTime>True</DesignTime>
       <DependentUpon>Levy learner changes employer negotiated total cost and ILR is submitted late - PV2-367.feature</DependentUpon>
     </Compile>
-    <Compile Include="Tests\Employer Change of Circumstances\Levy learner changes employer with change to negotiated price at the mid of month with gap- PV2-364.feature.cs">
```
[the remark this answers is not in the document]
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/SFA.DAS.Payments.AcceptanceTests.EndToEnd.csproj]: Noop the .cs file don't need to be in the solution just run custom tool to generate a new one from spec

File: src/SFA.DAS.Payments.Model.Core/Entities/ApprenticeshipPriceEpisodeModel.cs (hunk 3 above)
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.Model.Core/Entities/ApprenticeshipPriceEpisodeModel.cs]: This is how entity framework handles many-to -1 relationship.

File: src/SFA.DAS.Payments.Application/Repositories/PaymentsDataContext.cs (hunk 4 above)
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.Application/Repositories/PaymentsDataContext.cs]: Is been changed based on discussion with Tony. So Tony needs to update the spec.

File: src/SFA.DAS.Payments.Database/Payments2/Tables/Payments2.Apprenticeship.sql (hunk 6 above)
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.Database/Payments2/Tables/Payments2.Apprenticeship.sql]: Yes null means not set and  is nullable

File: src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/Steps/EndToEndStepsBase.cs (hunk 7 above)
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/Steps/EndToEndStepsBase.cs]: I believe  Apprenticeship is only active when they start.

File: src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/SFA.DAS.Payments.AcceptanceTests.EndToEnd.csproj [hunk 11]
```diff
@@ -1231,7 +1228,7 @@
     <Analyzer Include="..\packages\NServiceBus.7.1.5\analyzers\dotnet\cs\NServiceBus.Core.Analyzer.dll" />
   </ItemGroup>
   <ItemGroup>
-    <Folder Include="Tests\Levy Learner\Employer Change of Circumstances\" />
+    <Folder Include="Mapping\" />
```
[the remark this answers is not in the document]
Comment (samuel-dev-dba) [reply src/SFA.DAS.Payments.AcceptanceTests.EndToEnd/SFA.DAS.Payments.AcceptanceTests.EndToEnd.csproj]: Now removed

**Pull request merged**